src/main/java/org/spongepowered/common/entity/SpongeEntitySnapshot.java

```diff
@@ -0,0 +1,312 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.entity;
+
+import com.google.common.base.MoreObjects;
+import net.minecraft.nbt.CompoundNBT;
+import org.spongepowered.api.ResourceKey;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.data.persistence.DataContainer;
+import org.spongepowered.api.data.persistence.DataView;
+import org.spongepowered.api.data.persistence.InvalidDataException;
+import org.spongepowered.api.data.persistence.Queries;
+import org.spongepowered.api.entity.Entity;
+import org.spongepowered.api.entity.EntityArchetype;
+import org.spongepowered.api.entity.EntitySnapshot;
+import org.spongepowered.api.entity.EntityType;
+import org.spongepowered.api.event.CauseStackManager.StackFrame;
+import org.spongepowered.api.event.EventContextKeys;
+import org.spongepowered.api.event.cause.entity.SpawnTypes;
+import org.spongepowered.api.util.Transform;
+import org.spongepowered.api.world.ServerLocation;
+import org.spongepowered.api.world.server.ServerWorld;
+import org.spongepowered.common.SpongeCommon;
+import org.spongepowered.common.bridge.data.CustomDataHolderBridge;
+import org.spongepowered.common.bridge.data.DataContainerHolder;
+import org.spongepowered.common.bridge.world.storage.WorldInfoBridge;
+import org.spongepowered.common.data.holder.SpongeImmutableDataHolder;
+import org.spongepowered.common.data.persistence.NbtTranslator;
+import org.spongepowered.common.event.tracking.PhaseTracker;
+import org.spongepowered.common.util.Constants;
+import org.spongepowered.math.vector.Vector3d;
+import org.spongepowered.math.vector.Vector3i;
+
+import java.lang.ref.WeakReference;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+
+import javax.annotation.Nullable;
+
+public class SpongeEntitySnapshot implements EntitySnapshot, SpongeImmutableDataHolder<EntitySnapshot>, DataContainerHolder.Immutable<EntitySnapshot> {
+
+    @Nullable private final UUID entityUuid;
+    private final ResourceKey worldKey;
+    private final EntityType<?> entityType;
+    private final Vector3d position;
+    private final Vector3d rotation;
+    private final Vector3d scale;
+    @Nullable private final CompoundNBT compound;
+    @Nullable private final WeakReference<Entity> entityReference;
+
+    // for entities during events.
+    private boolean isDirty = true;
```

Comment:
eh? can you explain a little more?

Reply:
no can you? https://github.com/SpongePowered/SpongeCommon/commit/90407b053b2cf48439633264b2d411cefb296a97